pkg/virt-controller/watch/drain/disruptionbudget/disruptionbudget.go

```diff
@@ -527,45 +527,47 @@ func isPDBFromOldVMI(vmi *virtv1.VirtualMachineInstance, pdb *policyv1.PodDisrup
 
 func (c *DisruptionBudgetController) sync(key string, vmiExists bool, vmi *virtv1.VirtualMachineInstance, pdb *policyv1.PodDisruptionBudget) error {
 	needsEvictionProtection := c.vmiNeedsEvictionPDB(vmiExists, vmi)
-
-	// check for deletions if pod exists
-	if pdb != nil {
-		if !vmiExists || vmi.DeletionTimestamp != nil {
-			// being deleted
-			log.Log.Infof("deleting pdb %s/%s due to VMI deletion", pdb.Namespace, pdb.Name)
-			return c.deletePDB(key, pdb, vmi)
-		} else if !needsEvictionProtection {
-			// vmi isn't set to prevent eviction, so delete the pdb
-			log.Log.Object(vmi).Infof("deleting pdb %s/%s due to not using evictionStrategy: LiveMigration|External", pdb.Namespace, pdb.Name)
-			return c.deletePDB(key, pdb, vmi)
-		} else if isPDBFromOldVMI(vmi, pdb) {
-			// pdb for non existent vmi
-			log.Log.Object(vmi).Infof("deleting pdb %s/%s due to VMI not existing anymore", pdb.Namespace, pdb.Name)
-			return c.deletePDB(key, pdb, vmi)
-		} else if pdbs.IsPDBFromOldMigrationController(pdb) {
-			// pdb coming from an old migration controller
-			log.Log.Object(vmi).Infof("deleting pdb %s/%s generated by an old migration controller", pdb.Namespace, pdb.Name)
-			return c.deletePDB(key, pdb, vmi)
-		} else {
-			vmimCompleted, err := c.isVMIMCompletedForPDB(pdb, vmi)
-			if err != nil {
-				return err
-			}
-			if vmimCompleted {
-				// pdb for completed migration
-				log.Log.Object(vmi).Infof("shrinking pdb %s/%s due to migration completion", pdb.Namespace, pdb.Name)
-				return c.shrinkPDB(vmi, pdb)
-			}
+	if pdb == nil {
+		if needsEvictionProtection {
+			// If no PDB exists and eviction protection is needed, create a new PDB
+			log.Log.Object(vmi).Infof("creating pdb for VMI %s/%s", vmi.Namespace, vmi.Name)
+			return c.createPDB(key, vmi)
 		}
-	} else if needsEvictionProtection {
-		// pdb doesn't exist, create if vmi's eviction strategy means it is protected during drain.
-		log.Log.Object(vmi).Infof("creating pdb for VMI %s/%s", vmi.Namespace, vmi.Name)
-		return c.createPDB(key, vmi)
+		return nil
+	}
+
+	if shouldDelete, reason := shouldDeletePDB(vmiExists, vmi, pdb, needsEvictionProtection); shouldDelete {
+		log.Log.Object(vmi).Infof("deleting pdb %s/%s due to %s", pdb.Namespace, pdb.Name, reason)
+		return c.deletePDB(key, pdb, vmi)
+	}
+
+	vmimCompleted, err := c.isVMIMCompletedForPDB(pdb, vmi)
+	if err != nil {
+		return err
+	}
+	if vmimCompleted {
+		log.Log.Object(vmi).Infof("shrinking pdb %s/%s due to migration completion", pdb.Namespace, pdb.Name)
+		return c.shrinkPDB(vmi, pdb)
 	}
 
 	return nil
 }
 
+func shouldDeletePDB(vmiExists bool, vmi *virtv1.VirtualMachineInstance, pdb *policyv1.PodDisruptionBudget, needsEvictionProtection bool) (bool, string) {
```

Comment:
nit: In my opinion it would be valuable to use named return params here since the string parameter's purpose is not clear from looking at the signature:
```suggestion
func shouldDeletePDB(vmiExists bool, vmi *virtv1.VirtualMachineInstance, pdb *policyv1.PodDisruptionBudget, needsEvictionProtection bool) (shouldDelete bool, deletionReason string) {
```

Reply:
You're absolutely right; it looks much better now.